File: apps/android/app/src/main/java/com/apostolos/tv/data/FavoritesStore.kt
package com.apostolos.tv.data

import android.content.Context
import com.apostolos.tv.data.model.FavoriteItem
import com.apostolos.tv.data.model.FavoriteKind
import com.apostolos.tv.data.model.LiveStream
import com.apostolos.tv.data.model.SeriesItem
import com.apostolos.tv.data.model.VodStream
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

class FavoritesStore(
    context: Context,
    private val json: Json = Json { ignoreUnknownKeys = true },
) {
    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val _items = MutableStateFlow(load())
    val items: StateFlow<List<FavoriteItem>> = _items.asStateFlow()

    fun isFavorite(id: String): Boolean = _items.value.any { it.id == id }

    fun toggle(item: FavoriteItem): Boolean {
        val existing = _items.value.any { it.id == item.id }
        val updated = if (existing) {
            _items.value.filterNot { it.id == item.id }
        } else {
            listOf(item) + _items.value
        }
        _items.value = updated
        persist(updated)
        return !existing
    }

    fun itemsOf(kind: FavoriteKind): List<FavoriteItem> =
        _items.value.filter { it.kind == kind }

    fun toLiveStreams(items: List<FavoriteItem>): List<LiveStream> =
        items.filter { it.kind == FavoriteKind.LIVE }.map { fav ->
            LiveStream(
                name = fav.title,
                streamId = fav.streamId,
                streamIcon = fav.imageUrl,
                categoryId = fav.categoryId,
            )
        }

    fun toVodStreams(items: List<FavoriteItem>): List<VodStream> =
        items.filter { it.kind == FavoriteKind.MOVIE }.map { fav ->
            VodStream(
                name = fav.title,
                streamId = fav.streamId,
                streamIcon = fav.imageUrl,
                categoryId = fav.categoryId,
                containerExtension = fav.containerExtension,
            )
        }

    fun toSeriesItems(items: List<FavoriteItem>): List<SeriesItem> =
        items.filter { it.kind == FavoriteKind.SERIES }.map { fav ->
            SeriesItem(
                name = fav.title,
                seriesId = fav.seriesId,
                cover = fav.imageUrl,
                categoryId = fav.categoryId,
            )
        }

    fun clear() {
        _items.value = emptyList()
        prefs.edit().remove(KEY_ITEMS).apply()
    }

    fun replaceAll(items: List<FavoriteItem>) {
        _items.value = items
        persist(items)
    }

    private fun load(): List<FavoriteItem> {
        val payload = prefs.getString(KEY_ITEMS, null) ?: return emptyList()
        return runCatching { json.decodeFromString<List<FavoriteItem>>(payload) }
            .getOrElse { emptyList() }
    }

    private fun persist(items: List<FavoriteItem>) {
        prefs.edit()
            .putString(KEY_ITEMS, json.encodeToString(items))
            .apply()
    }

    companion object {
        private const val PREFS_NAME = "tv_favorites"
        private const val KEY_ITEMS = "items"
    }
}

File: apps/tizen-tv/src/data/favoritesStore.ts
export type FavoriteKind = "LIVE" | "MOVIE" | "SERIES";

export interface FavoriteItem {
  id: string;
  kind: FavoriteKind;
  title: string;
  imageUrl?: string;
  streamId?: number;
  seriesId?: number;
  containerExtension?: string;
  categoryId?: string;
}

const STORAGE_KEY = "tv_favorites";

export class FavoritesStore {
  private items: FavoriteItem[] = load();

  getAll(): FavoriteItem[] {
    return [...this.items];
  }

  isFavorite(id: string): boolean {
    return this.items.some((item) => item.id === id);
  }

  toggle(item: FavoriteItem): boolean {
    const exists = this.isFavorite(item.id);
    this.items = exists
      ? this.items.filter((entry) => entry.id !== item.id)
      : [item, ...this.items];
    persist(this.items);
    return !exists;
  }

  replaceAll(items: FavoriteItem[]): void {
    this.items = items;
    persist(this.items);
  }

  byKind(kind: FavoriteKind): FavoriteItem[] {
    return this.items.filter((item) => item.kind === kind);
  }
}

export function favoriteFromLive(stream: {
  stream_id: number;
  name: string;
  stream_icon?: string;
  category_id?: string;
}): FavoriteItem {
  return {
    id: `fav_live_${stream.stream_id}`,
    kind: "LIVE",
    title: stream.name,
    imageUrl: stream.stream_icon,
    streamId: stream.stream_id,
    categoryId: stream.category_id,
  };
}

export function favoriteFromMovie(movie: {
  stream_id: number;
  name: string;
  stream_icon?: string;
  category_id?: string;
  container_extension?: string;
}): FavoriteItem {
  return {
    id: `fav_movie_${movie.stream_id}`,
    kind: "MOVIE",
    title: movie.name,
    imageUrl: movie.stream_icon,
    streamId: movie.stream_id,
    categoryId: movie.category_id,
    containerExtension: movie.container_extension,
  };
}

export function favoriteFromSeries(series: {
  series_id: number;
  name: string;
  cover?: string;
  category_id?: string;
}): FavoriteItem {
  return {
    id: `fav_series_${series.series_id}`,
    kind: "SERIES",
    title: series.name,
    imageUrl: series.cover,
    seriesId: series.series_id,
    categoryId: series.category_id,
  };
}

function load(): FavoriteItem[] {
  try {
    const raw = localStorage.getItem(STORAGE_KEY);
    return raw ? (JSON.parse(raw) as FavoriteItem[]) : [];
  } catch {
    return [];
  }
}

function persist(items: FavoriteItem[]): void {
  localStorage.setItem(STORAGE_KEY, JSON.stringify(items));
}

export const favoritesStore = new FavoritesStore();
